File: basic/15-k-means/k-means.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define MAX_POINTS 100
#define MAX_CLUSTERS 10

typedef struct {
    double x, y;
} Point;

typedef struct {
    Point centroid;
    Point points[MAX_POINTS];
    int point_count;
} Cluster;

double distance(Point a, Point b) {
    return sqrt(pow(a.x - b.x, 2) + pow(a.y - b.y, 2));
}

void kmeans(Point points[], int num_points, Cluster clusters[], int k) {
    // 初始化聚类中心
    for (int i = 0; i < k; i++) {
        clusters[i].centroid = points[i]; // 随机选择初始中心
        clusters[i].point_count = 0;
    }

    int changed;
    do {
        changed = 0;

        // 清空之前的点
        for (int i = 0; i < k; i++) {
            clusters[i].point_count = 0;
        }

        // 分配每个点到最近的聚类
        for (int i = 0; i < num_points; i++) {
            int best_cluster = 0;
            double min_dist = distance(points[i], clusters[0].centroid);

            for (int j = 1; j < k; j++) {
                double dist = distance(points[i], clusters[j].centroid);
                if (dist < min_dist) {
                    min_dist = dist;
                    best_cluster = j;
                }
            }
            clusters[best_cluster].points[clusters[best_cluster].point_count++] = points[i];
            changed = 1;
        }

        // 更新聚类中心
        for (int i = 0; i < k; i++) {
            double sum_x = 0, sum_y = 0;
            for (int j = 0; j < clusters[i].point_count; j++) {
                sum_x += clusters[i].points[j].x;
                sum_y += clusters[i].points[j].y;
            }
            if (clusters[i].point_count > 0) {
                clusters[i].centroid.x = sum_x / clusters[i].point_count;
                clusters[i].centroid.y = sum_y / clusters[i].point_count;
            }
        }
    } while (changed);
}

int main() {
    Point points[MAX_POINTS] = {{1, 2}, {1, 4}, {1, 0}, {10, 2}, {10, 4}, {10, 0}};
    int num_points = 6;
    int k = 2;
    Cluster clusters[MAX_CLUSTERS];

    kmeans(points, num_points, clusters, k);

    for (int i = 0; i < k; i++) {
        printf("Cluster %d centroid: (%.2f, %.2f)\n", i, clusters[i].centroid.x, clusters[i].centroid.y);
    }

    return 0;
}

File: basic/15-k-means/k-means.java
import java.util.Arrays;
import java.util.Random;

public class KMeans {
    static class Point {
        double x, y;

        Point(double x, double y) {
            this.x = x;
            this.y = y;
        }
    }

    static double distance(Point a, Point b) {
        return Math.sqrt(Math.pow(a.x - b.x, 2) + Math.pow(a.y - b.y, 2));
    }

    static void kmeans(Point[] points, int k) {
        Point[] centroids = new Point[k];
        Random rand = new Random();

        // 随机选择初始聚类中心
        for (int i = 0; i < k; i++) {
            centroids[i] = points[rand.nextInt(points.length)];
        }

        boolean changed;
        do {
            changed = false;
            Point[][] clusters = new Point[k][points.length];
            int[] counts = new int[k];

            // 分配每个点到最近的聚类中心
            for (Point point : points) {
                int bestCluster = 0;
                double minDist = distance(point, centroids[0]);

                for (int j = 1; j < k; j++) {
                    double dist = distance(point, centroids[j]);
                    if (dist < minDist) {
                        minDist = dist;
                        bestCluster = j;
                    }
                }
                clusters[bestCluster][counts[bestCluster]++] = point;
            }

            // 更新聚类中心
            for (int i = 0; i < k; i++) {
                double sumX = 0, sumY = 0;
                for (int j = 0; j < counts[i]; j++) {
                    sumX += clusters[i][j].x;
                    sumY += clusters[i][j].y;
                }
                if (counts[i] > 0) {
                    Point newCentroid = new Point(sumX / counts[i], sumY / counts[i]);
                    if (distance(centroids[i], newCentroid) > 1e-5) {
                        changed = true;
                    }
                    centroids[i] = newCentroid;
                }
            }
        } while (changed);

        for (int i = 0; i < k; i++) {
            System.out.printf("Cluster %d centroid: (%.2f, %.2f)\n", i, centroids[i].x, centroids[i].y);
        }
    }

    public static void main(String[] args) {
        Point[] points = {new Point(1, 2), new Point(1, 4), new Point(1, 0), new Point(10, 2), new Point(10, 4), new Point(10, 0)};
        int k = 2;
        kmeans(points, k);
    }
}
